RecoLocalTracker/SiPhase2VectorHitBuilder/src/VectorHitBuilderAlgorithm.cc

```diff
@@ -0,0 +1,553 @@
+#include "RecoLocalTracker/SiPhase2VectorHitBuilder/interface/VectorHitBuilderAlgorithm.h"
+#include "RecoLocalTracker/ClusterParameterEstimator/interface/ClusterParameterEstimator.h"
+#include "Geometry/CommonDetUnit/interface/GeomDet.h"
+#include "Geometry/CommonTopologies/interface/PixelTopology.h"
+#include "DataFormats/TrackerRecHit2D/interface/VectorHit2D.h"
+
+void VectorHitBuilderAlgorithm::run(edm::Handle<edmNew::DetSetVector<Phase2TrackerCluster1D>> clusters,
+                                    VectorHitCollection& vhAcc,
+                                    VectorHitCollection& vhRej,
+                                    edmNew::DetSetVector<Phase2TrackerCluster1D>& clustersAcc,
+                                    edmNew::DetSetVector<Phase2TrackerCluster1D>& clustersRej) const {
+  LogDebug("VectorHitBuilderAlgorithm") << "Run VectorHitBuilderAlgorithm ... \n";
+  const auto* clustersPhase2Collection = clusters.product();
+
+  std::unordered_map<DetId, std::vector<VectorHit>> tempVHAcc, tempVHRej;
+  int totalAccepted = 0;
+  int totalRejected = 0;
+
+  //loop over the DetSetVector
+  LogDebug("VectorHitBuilderAlgorithm") << "with #clusters : " << clustersPhase2Collection->size() << std::endl;
+  for (auto dSViter : *clustersPhase2Collection) {
+    unsigned int rawDetId1(dSViter.detId());
+    DetId detId1(rawDetId1);
+    DetId lowerDetId, upperDetId;
+    if (tkTopo_->isLower(detId1)) {
+      lowerDetId = detId1;
+      upperDetId = tkTopo_->partnerDetId(detId1);
+    } else
+      continue;
+
+    DetId detIdStack = tkTopo_->stack(detId1);
+
+    //debug
+    LogDebug("VectorHitBuilderAlgorithm") << "  DetId stack : " << detIdStack.rawId() << std::endl;
+    LogDebug("VectorHitBuilderAlgorithm") << "  DetId lower set of clusters  : " << lowerDetId.rawId();
+    LogDebug("VectorHitBuilderAlgorithm") << "  DetId upper set of clusters  : " << upperDetId.rawId() << std::endl;
+
+    const GeomDet* gd;
+    const StackGeomDet* stackDet;
+    const auto& it_detLower = dSViter;
+    const auto& it_detUpper = clustersPhase2Collection->find(upperDetId);
+
+    if (it_detUpper != clustersPhase2Collection->end()) {
+      gd = tkGeom_->idToDet(detIdStack);
+      stackDet = dynamic_cast<const StackGeomDet*>(gd);
+      std::vector<VectorHit> vhsInStack_Acc;
+      std::vector<VectorHit> vhsInStack_Rej;
+      const auto& vhsInStack_AccRej = buildVectorHits(stackDet, clusters, it_detLower, *it_detUpper);
+
+      //storing accepted and rejected VHs
+      for (const auto& vh : vhsInStack_AccRej) {
+        if (vh.second == true) {
+          vhsInStack_Acc.push_back(vh.first);
+          std::push_heap(vhsInStack_Acc.begin(), vhsInStack_Acc.end());
+          totalAccepted += 1;
+        } else if (vh.second == false) {
+          vhsInStack_Rej.push_back(vh.first);
+          totalRejected += 1;
```

Comment:
what's the purpose of sorting the hits (by chi2) in the accepted list but not in the rejected?
Is this something special encoded for seeding only?

Reply:
If not for this sorting, I'd suggest to get rid of `tempVHAcc` and `tempVHRej` maps and use `FastFiller` directly,
possibly even inside the `buildVectorHits`
The stacks are supposed to be ordered in the same way as we read the data.
Making first 2M hits in 13K vectors just to push them to the final 1D vector in the same order seems wasteful.